extensions/listeners/eth_deposits/ethereum.go

```diff
@@ -241,6 +218,10 @@ func retry(maxRetries int64, fn func() error) error {
 			return err
 		}
 
-		time.Sleep(retrier.Duration())
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(retrier.Duration()):
```

Comment:
I don't fully understand the leaking timer issue, but do we have this problem here?

I'm assuming no unless there is the very occasional context cancellation, which is NBD, but I figured I'd ask since I'm not sure.

Reply:
A timer leak is of concern in the following general situation:

```
	for {
		select {
		case <-fastSignal:
		case <-time.After(longTime): // always creates ticker, rarely received from
		}
	}
```

Each iteration of the loop will create a new ticker internally just for the `time.After(longTime)` call before receiving on any `case`.  When it's rare to receive on that `case`, but you hit the `select` a lot, then those tickers aren't garbage collected until they fire (later when while nothing is listening).  This is what https://pkg.go.dev/time#After means by:

> The underlying Timer is not recovered by the garbage collector until the timer fires. If efficiency is a concern, use NewTimer instead and call Timer.Stop if the timer is no longer needed.

When a timeout is added in such a select in a loop that normally receives quickly, you can leak a ton of timers, even crash the machine.  The general solution is to use a timer or ticker with a deferred `Stop` so that gc can clean it up immediately after.

In the case case of this `retry` body, a context cancel means you aren't going to retry again, so you only leak **1** timer, which is if little consequence particularly because the listener or entire app is shutting down.

Reply:
Makes total sense, thanks for explaining that.